File: frontend/src/api/customContent.ts
import { apiDelete, apiGet, apiPost, apiPut } from './client';

// The 7 custom entity types (DHCM-20/28), each global/instance-wide with no
// shared shape -- mirrors backend/app/routers/custom_content.py's segments.
export type CustomContentSegment =
  | 'classes'
  | 'ancestries'
  | 'communities'
  | 'domains'
  | 'domain-cards'
  | 'weapons'
  | 'armor';

export type CustomFieldType = 'text' | 'number' | 'boolean' | 'textarea';

export interface CustomFieldConfig {
  name: string;
  label: string;
  type: CustomFieldType;
  required?: boolean;
  defaultValue?: string;
}

export const SEGMENT_LABELS: Record<CustomContentSegment, string> = {
  classes: 'Classes',
  ancestries: 'Ancestries',
  communities: 'Communities',
  domains: 'Domains',
  'domain-cards': 'Domain Cards',
  weapons: 'Weapons',
  armor: 'Armor',
};

// Armor has no plural to strip ("Armor", not "Armors"), so a shared regex
// can't derive these from SEGMENT_LABELS -- spelled out explicitly instead.
export const SEGMENT_SINGULAR_LABELS: Record<CustomContentSegment, string> = {
  classes: 'Class',
  ancestries: 'Ancestry',
  communities: 'Community',
  domains: 'Domain',
  'domain-cards': 'Domain Card',
  weapons: 'Weapon',
  armor: 'Armor',
};

// Field order/type per segment, matching each Create schema in
// backend/app/schemas/custom_content.py 1:1 -- drives the generic form
// instead of hand-building 7 bespoke ones.
export const SEGMENT_FIELDS: Record<CustomContentSegment, CustomFieldConfig[]> = {
  classes: [
    { name: 'name', label: 'Name', type: 'text', required: true },
    { name: 'domains_json', label: 'Domains (JSON list)', type: 'textarea', required: true },
    { name: 'starting_evasion', label: 'Starting Evasion', type: 'number', required: true },
    { name: 'starting_hp', label: 'Starting HP', type: 'number', required: true },
    { name: 'class_items_json', label: 'Class Items (JSON list)', type: 'textarea', required: true },
    { name: 'subclasses_json', label: 'Subclasses (JSON list)', type: 'textarea', required: true },
  ],
  ancestries: [
    { name: 'name', label: 'Name', type: 'text', required: true },
    { name: 'features_json', label: 'Features (JSON list)', type: 'textarea', defaultValue: '[]' },
  ],
  communities: [
    { name: 'name', label: 'Name', type: 'text', required: true },
    {
      name: 'adjectives_json',
      label: 'Adjectives (JSON list)',
      type: 'textarea',
      defaultValue: '[]',
    },
    {
      name: 'feature_json',
      label: 'Feature (JSON object or null)',
      type: 'textarea',
      defaultValue: 'null',
    },
  ],
  domains: [
    { name: 'name', label: 'Name', type: 'text', required: true },
    { name: 'classes_json', label: 'Classes (JSON list)', type: 'textarea', required: true },
  ],
  'domain-cards': [
    { name: 'domain', label: 'Domain', type: 'text', required: true },
    { name: 'name', label: 'Name', type: 'text', required: true },
    { name: 'type', label: 'Type', type: 'text', required: true },
    { name: 'recall_cost', label: 'Recall Cost', type: 'number', required: true },
  ],
  weapons: [
    { name: 'name', label: 'Name', type: 'text', required: true },
    { name: 'trait', label: 'Trait', type: 'text', required: true },
    { name: 'range', label: 'Range', type: 'text', required: true },
    { name: 'damage', label: 'Damage', type: 'text', required: true },
    { name: 'burden', label: 'Burden', type: 'text', required: true },
    { name: 'is_magic', label: 'Magic', type: 'boolean' },
    { name: 'feature', label: 'Feature (optional)', type: 'textarea' },
  ],
  armor: [
    { name: 'name', label: 'Name', type: 'text', required: true },
    { name: 'threshold_low', label: 'Threshold (Low)', type: 'number', required: true },
    { name: 'threshold_high', label: 'Threshold (High)', type: 'number', required: true },
    { name: 'base_score', label: 'Base Score', type: 'number', required: true },
    { name: 'feature', label: 'Feature (optional)', type: 'textarea' },
  ],
};

export interface CustomEntity {
  id: number;
  name: string;
  created_at: string;
  [key: string]: unknown;
}

function url(segment: CustomContentSegment, entityId?: number): string {
  const base = `/api/custom-content/${segment}`;
  return entityId === undefined ? base : `${base}/${entityId}`;
}

export function listCustomEntities(segment: CustomContentSegment): Promise<CustomEntity[]> {
  return apiGet(url(segment));
}

export function createCustomEntity(
  segment: CustomContentSegment,
  body: Record<string, unknown>,
): Promise<CustomEntity> {
  return apiPost(url(segment), body);
}

export function updateCustomEntity(
  segment: CustomContentSegment,
  entityId: number,
  body: Record<string, unknown>,
): Promise<CustomEntity> {
  return apiPut(url(segment, entityId), body);
}

export function deleteCustomEntity(
  segment: CustomContentSegment,
  entityId: number,
): Promise<void> {
  return apiDelete(url(segment, entityId));
}

File: backend/app/schemas/custom_content.py
"""Custom-content schemas: GM-authored classes/ancestries/communities/domains/
domain cards/weapons/armor, alongside the SRD's static datasets (DHCM-20).

Each type has its own shape (matching its model's columns 1:1), so each gets
its own Create/Update/Out trio rather than a shared generic schema.
"""

from datetime import datetime

from pydantic import BaseModel, Field


class CustomClassCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    domains_json: str
    starting_evasion: int
    starting_hp: int
    class_items_json: str
    subclasses_json: str


class CustomClassUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    domains_json: str | None = None
    starting_evasion: int | None = None
    starting_hp: int | None = None
    class_items_json: str | None = None
    subclasses_json: str | None = None

    model_config = {"extra": "forbid"}


class CustomClassOut(BaseModel):
    id: int
    name: str
    domains_json: str
    starting_evasion: int
    starting_hp: int
    class_items_json: str
    subclasses_json: str
    created_at: datetime


class CustomAncestryCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    features_json: str = "[]"


class CustomAncestryUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    features_json: str | None = None

    model_config = {"extra": "forbid"}


class CustomAncestryOut(BaseModel):
    id: int
    name: str
    features_json: str
    created_at: datetime


class CustomCommunityCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    adjectives_json: str = "[]"
    feature_json: str = "null"


class CustomCommunityUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    adjectives_json: str | None = None
    feature_json: str | None = None

    model_config = {"extra": "forbid"}


class CustomCommunityOut(BaseModel):
    id: int
    name: str
    adjectives_json: str
    feature_json: str
    created_at: datetime


class CustomDomainCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    classes_json: str


class CustomDomainUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    classes_json: str | None = None

    model_config = {"extra": "forbid"}


class CustomDomainOut(BaseModel):
    id: int
    name: str
    classes_json: str
    created_at: datetime


class CustomDomainCardCreate(BaseModel):
    domain: str = Field(min_length=1, max_length=100)
    name: str = Field(min_length=1, max_length=100)
    type: str = Field(min_length=1, max_length=50)
    recall_cost: int


class CustomDomainCardUpdate(BaseModel):
    domain: str | None = Field(default=None, min_length=1, max_length=100)
    name: str | None = Field(default=None, min_length=1, max_length=100)
    type: str | None = Field(default=None, min_length=1, max_length=50)
    recall_cost: int | None = None

    model_config = {"extra": "forbid"}


class CustomDomainCardOut(BaseModel):
    id: int
    domain: str
    name: str
    type: str
    recall_cost: int
    created_at: datetime


class CustomWeaponCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    trait: str = Field(min_length=1, max_length=20)
    range: str = Field(min_length=1, max_length=20)
    damage: str = Field(min_length=1, max_length=50)
    burden: str = Field(min_length=1, max_length=20)
    is_magic: bool
    feature: str | None = None


class CustomWeaponUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    trait: str | None = Field(default=None, min_length=1, max_length=20)
    range: str | None = Field(default=None, min_length=1, max_length=20)
    damage: str | None = Field(default=None, min_length=1, max_length=50)
    burden: str | None = Field(default=None, min_length=1, max_length=20)
    is_magic: bool | None = None
    feature: str | None = None

    model_config = {"extra": "forbid"}


class CustomWeaponOut(BaseModel):
    id: int
    name: str
    trait: str
    range: str
    damage: str
    burden: str
    is_magic: bool
    feature: str | None
    created_at: datetime


class CustomArmorCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    threshold_low: int
    threshold_high: int
    base_score: int
    feature: str | None = None


class CustomArmorUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    threshold_low: int | None = None
    threshold_high: int | None = None
    base_score: int | None = None
    feature: str | None = None

    model_config = {"extra": "forbid"}


class CustomArmorOut(BaseModel):
    id: int
    name: str
    threshold_low: int
    threshold_high: int
    base_score: int
    feature: str | None
    created_at: datetime
